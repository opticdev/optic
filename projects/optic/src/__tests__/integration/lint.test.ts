import {
  test,
  expect,
  describe,
  jest,
  beforeEach,
  afterEach,
} from '@jest/globals';
import { runOptic, setupWorkspace, normalizeWorkspace } from './integration';

jest.setTimeout(30000);

let oldEnv: any;
beforeEach(() => {
  oldEnv = { ...process.env };
  process.env.LOG_LEVEL = 'info';
  process.env.OPTIC_ENV = 'local';
  process.env.CI = 'false';
});

afterEach(() => {
  process.env = { ...oldEnv };
});

describe('lint', () => {
  test('fails on validation errors', async () => {
    const workspace = await setupWorkspace('lint/specs');
    const { combined, code } = await runOptic(
      workspace,
      'lint spec-fails-validation.yml'
    );
    expect(normalizeWorkspace(workspace, combined)).toMatchSnapshot();
    expect(code).toBe(1);
  });

  test('fails on requirement rule errors', async () => {
    const workspace = await setupWorkspace('lint/specs');
    const { combined, code } = await runOptic(
      workspace,
      'lint spec-fails-requirement.yml'
    );
    expect(normalizeWorkspace(workspace, combined)).toMatchSnapshot();
    expect(code).toBe(1);
  });

  test('fails on bad formatting', async () => {
    const workspace = await setupWorkspace('lint/specs');
    const { combined, code } = await runOptic(
      workspace,
      'lint spec-with-bad-formatting.yml'
    );
    expect(normalizeWorkspace(workspace, combined)).toMatchSnapshot();
    expect(code).toBe(1);
  });

  test('passes on valid specs', async () => {
    const workspace = await setupWorkspace('lint/specs');
    const { combined, code } = await runOptic(
      workspace,
      'lint spec-good-spec.yml'
    );
    expect(normalizeWorkspace(workspace, combined)).toMatchSnapshot();
    expect(code).toBe(0);
  });
});
