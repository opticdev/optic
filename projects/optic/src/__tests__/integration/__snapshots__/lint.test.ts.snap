// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lint fails on bad formatting 1`] = `
"Linting spec spec-with-bad-formatting.yml...
[31minvalid openapi: [39m[1m[31mpaths > /filler_route > post > responses > 201 > content > application/json > schema > properties > example must be object[39m[22m
[90m18 |[39m                   id:
[90m19 |[39m                     type: string
[90m20 |[39m                   format: uuid
[90m21 |[39m [1m[31m                  example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2[39m[22m
[90m22 |[39m 
[90m$workspace$/spec-with-bad-formatting.yml[39m
[31minvalid openapi: [39m[1m[31mpaths > /filler_route > post > responses > 201 > content > application/json > schema > properties > format must be object[39m[22m
[90m17 |[39m                 properties:
[90m18 |[39m                   id:
[90m19 |[39m                     type: string
[90m20 |[39m [1m[31m                  format: uuid[39m[22m
[90m21 |[39m                   example: d5b640e5-d88c-4c17-9bf0-93597b7a1ce2
[90m22 |[39m 
[90m$workspace$/spec-with-bad-formatting.yml[39m
"
`;

exports[`lint fails on requirement rule errors 1`] = `
"Linting spec spec-fails-requirement.yml...

Checks

  [1m[41m[37m FAIL [39m[49m[22m [1mPOST[22m /filler_route
    requirement rule: response property naming check
      [31m[31mx[39m[31m id_thing is not camelCase[39m
      at [4m$workspace$/spec-fails-requirement.yml:18:359[24m




[32m[1m0 checks passed[22m[39m
[31m[1m1 checks failed[22m[39m

[31m[1mLinting errors found with your OpenAPI spec.[22m[39m
"
`;

exports[`lint fails on validation errors 1`] = `
"Linting spec spec-fails-validation.yml...
[31minvalid openapi: [39m[1m[31minfo must have required property 'title'[39m[22m
[90m1 |[39m openapi: 3.0.3
[90m2 |[39m [1m[31minfo:[39m[22m
[90m3 |[39m   version: 0.1.0
[90m4 |[39m paths:
[90m5 |[39m   /filler_route:
[90m$workspace$/spec-fails-validation.yml[39m
"
`;

exports[`lint passes on valid specs 1`] = `
"Linting spec spec-good-spec.yml...

Checks

  [1m[42m[37m PASS [39m[49m[22m [1mPOST[22m /filler_route



[32m[1m1 checks passed[22m[39m
[31m[1m0 checks failed[22m[39m

[32m[1mLinting passed.[22m[39m
"
`;
